### PR TITLE
Add implementation

### PR DESCRIPTION
**Pull request by juliangruber**

### CODE REVIEW

File: index.js
Comment (bajtos) [inline index.js:4]: Nitpick: class names should be nouns.

E.g.

```suggestion
export class StuckTransactionsCanceller {
```

File: README.md
Comment (bajtos) [inline README.md:29]: I find this a bit difficult to comprehend. Either these three methods should be grouped into a "store" object/interface, or else their names should indicate they work with transactions.

Comment (bajtos) [inline README.md:29]: I find it surprising that a function called `resolve` deletes the data.

Considering that the first method is called "store", I propose to rename "resolve" to something like "drop" or "remove".

File: index.js
Comment (bajtos) [inline index.js:34]: I find the naming of these methods confusing.

Some are dealing with the status of individual transactions (pending, successful), this one performs a scan + replace action, which is not clear from the description.

Methods & functions usually start with a verb.

Proposed names:

```
addPending(tx)
removeSuccessful(tx)
cancelAllOlderThan(ageMs)
```

File: index.js
Comment (bajtos) [inline index.js:40]: Nitpick & micro-optimisation, feel free to ingore.

```suggestion
      return Date.now() - tx.timestamp > ageMs
```

File: index.js
Comment (bajtos) [inline index.js:50]: Please include the nonce in the log message. I found it very useful while cancelling the >400 stuck messages.

File: index.js
Comment (bajtos) [inline index.js:91]: Please add error handling.

- `sendTransaction` can fail with `err.code === 'NONCE_EXPIRED'` if we are too late and the transaction was added to a recently mined block
- `replacementTx.wait()` throws an error if the TX was rejected, I don't have a good example to show, but I encountered this while running my auto-cancellation script

It would be great to extract `cancelTx` and `getRecentSendMessage` into a public function exported by this module, so that I can re-use it in my script.

https://github.com/filecoin-station/utils/pull/1

File: test.js
Comment (bajtos) [inline test.js:85]: I have no idea what this test does, except that it seems to test everything.

Could you please break it down into multiple smaller and focused tests?

This can be done after the pull request is landed.

**Review by bajtos — APPROVED** [review]
Great work! I can tell this was created hastily. I found quite a few things that could be improved, but none are blocking this pull request from landing.

File: package.json
Comment (bajtos) [inline package.json:2]: Should we use a scoped package name? 

```suggestion
  "name": "@filecoin-station/cancel-stuck-transactions",
```

File: README.md
Comment (juliangruber) [reply README.md:29]: They aren't working with transactions: It's not a full tx object from ethers, and it has more keys at the same time.

I've been thinking about grouping these functions in a store object, will do that then 👍

File: index.js
Comment (juliangruber) [reply index.js:40]: I dislike this optimization, subtracting dates is more semantic than subtracting numbers

File: package.json
Comment (juliangruber) [reply package.json:2]: I had it scoped before but realized it has nothing to do with Station. If anything, it should be filecoin scoped

File: index.js
Comment (juliangruber) [reply index.js]: I was writing this "magic" style where it reads more like spoken word. I've changed it to be explicit

**Comment by bajtos** [comment]
@juliangruber please re-request another review when the PR is ready

File: index.js
Comment (juliangruber) [reply index.js]: > replacementTx.wait() throws an error if the TX was rejected, I don't have a good example to show, but I encountered this while running my auto-cancellation script

What kind of error handling should I add for this though? 🤔

File: package.json
Comment (juliangruber) [reply package.json:2]: > OTOH, in the past, we liked the convention that npm package name is the same as the github org/repo slug, that's why I suggested this change.

I like the name to fit the repo name but I personally don't care which org it's in.

> Maybe https://github.com/space-meridian would be a better home & npm scope for this package?

You're right, it also doesn't fit in this org really. I think it would be nice to keep it next to the `spark-*` repos, because that's where it's needed the most.

File: index.js
Comment (juliangruber) [reply index.js]: > I guess we have to treat this error as a generic "unknown error" and let the caller figure out what to do about it (e.g. report the problem to Sentry).

That's how I was thinking about it as well 👍 

> Is it worth mentioning in the docs that cancelOlderThan() will throw if the network rejects some of the replacement transactions?

Since we don't know yet how it can throw, I'd advertise for letting it through and then documenting these errors.

I'll also document the known nonce error next.

> Also, since we are cancelling multiple TXs concurrently, we need to ensure the error allows us to link the error details to the hash of the TX we were waiting to get confirmed. Linking to both the replaced and the replacement TXs will be even better.

When you run `cancelOlderThan()` it does a `Promise.all()` of canceling all transactions. The first one to reject will create the error that is thrown. I don't know yet whether the tx will be attached already. Are you ok with letting it fail and then seeing how to best improve?

Comment (bajtos) [reply index.js]: > > Also, since we are cancelling multiple TXs concurrently, we need to ensure the error allows us to link the error details to the hash of the TX we were waiting to get confirmed. Linking to both the replaced and the replacement TXs will be even better.
> 
> When you run `cancelOlderThan()` it does a `Promise.all()` of canceling all transactions. The first one to reject will create the error that is thrown.

Hmm, what happens with the other concurrent tasks when one of them fails and `Promise.all` returns? Will the tasks continue to completion until the replacement TX is submitted and confirmed? What happens if the caller calls `cancelOlderThan` again before those tasks are complete? Can we end up with multiple replacement transactions for the same stuck transaction?

Maybe we should use `Promise.allSettled` instead.

> I don't know yet whether the tx will be attached already. Are you ok with letting it fail and then seeing how to best improve?

Sure 👍🏻

File: README.md
Comment (bajtos) [inline README.md:16]: Nitpick - feel free to ignore.

Conceptually, `store` is a key-value store using `hash` as the key. I think the following API would make this fact more obvious to users:

```
async add (hash, { timestamp, from, gasPremium, nonce }) {
```

What is the semantics of this method when it comes to repeatedly adding the same key? In other words, what is the expected outcome of the following code:

```js
store.add({hash: 'hash', nonce: 10, /* other data */ })
store.add({hash: 'hash', nonce: 20, /* other data */ })
```

I think `store.set` is a better name. (Other alternatives that come to my mind: `store.store`, `store.save`.)

We could take it one step further and use the same method names as the built-in `Map` object provides:

 - `set(hash, data)`
 - `entries()`
 - `delete(hash)`

File: README.md
Comment (bajtos) [inline README.md:59]: The transaction does not necessarilly must be successful, right? If the smart contract rejects the method invocation in the TX, the transaction is still confirmed and added to the blockchain, and therefore we should remove it from the list of pending transactions.

Proposed name: `removeConfirmed` or `removeCommitted`.

File: README.md
Comment (bajtos) [inline README.md:83]: Since the function works with a single transaction, I think the name should also use a singular form.

```suggestion
- `sendTransaction`: `(Transaction) -> Promise<Transaction>`
```

File: index.js
Comment (juliangruber) [reply index.js]: > Maybe we should use Promise.allSettled instead.

I was thinking about this as well. In the old implementation (before this lib), individual cancel failures would simply be logged (and retried later) but the overall operation would succeed. I'll do as advised

File: index.js
Comment (bajtos) [inline index.js:30]: It would be great to add JSDoc comments to describe the return type of this function to enable IDE auto-completion for users of this module. It's fine to do that in a follow-up PR.

File: test.js
Comment (bajtos) [inline test.js:44]: I have good experience with using `debug` for test logs. Then, I can run `DEBUG=test node --test` to get verbose logs from the tests.

File: index.js
Comment (juliangruber) [reply index.js]: https://github.com/filecoin-station/cancel-stuck-transactions/pull/1/commits/b24f50a3623306f848c38c4aaf9957684bb1232f

**Review by bajtos — COMMENTED** [review]
Almost there!

File: README.md
Comment (juliangruber) [reply README.md:16]: I didn't think about this too much because it's impossible for two transactions to have the same hash. You're right that `set()` is more correct, but then I think we should rename `addPending(tx)` to `setPending(tx)` as well. Do you think this is worth it now?

> We could take it one step further and use the same method names as the built-in Map object provides:

This could be nice if you wanted to pass `{ store: new Map() }`. I think this would be nice but can be added later on.

File: README.md
Comment (juliangruber) [reply README.md:83]: This was only a docs typo :)

File: index.js
Comment (juliangruber) [reply index.js:30]: I'm happy to accept PRs :)

Comment (juliangruber) [reply index.js:30]: I personally don't use features like this

File: test.js
Comment (juliangruber) [reply test.js:47]: Great idea!

File: README.md
Comment (bajtos) [reply README.md:16]: > I didn't think about this too much because it's impossible for two transactions to have the same hash. You're right that `set()` is more correct, 

I think it's worth improving the `store` API to make it obvious to implementers what the expected semantics are. It's not very clear to me that the `store` API assumes that "it's impossible for two transactions to have the same hash", I see the store API more of a generic key-value store. But maybe that's just me.

> but then I think we should rename `addPending(tx)` to `setPending(tx)` as well. 

I am fine with `addPending`, since at this level of abstraction, we are dealing with full transaction objects where it's obvious that adding the same TX second time is a no-op.

> Do you think this is worth it now?

Yes, it's a quick change that only affects this PR. If we leave the change until later, we would have to rework all consumers of this module, which means more work.
 
Having said that, we are probably spending more time on discussing this change than it's worth it. If you are fine with `store.add`, then let's keep it.

> > We could take it one step further and use the same method names as the built-in Map object provides:
> 
> This could be nice if you wanted to pass `{ store: new Map() }`. I think this would be nice but can be added later on.

The benefit of using a Map-like API is that developers don't have to learn a new shape of a key/value store API, they can use what they are familiar with.

Having said that, this is a minor detail, let's not bikeshed.

File: test.js
Comment (bajtos) [inline test.js:47]: Are the comments `// TODO: Test logs` still relevant? Can we remove them now?

File: test.js
Comment (bajtos) [inline test.js:49]: Nitpick, feel free to ignore.

```suggestion
      log: str => debug(str)
```

(This also applies to all tests below.)

**Review by bajtos — APPROVED** [review]
LGTM with a few minor comments to consider.

File: README.md
Comment (juliangruber) [reply README.md:16]: Ok I'll go with `add()`

File: test.js
Comment (juliangruber) [reply test.js:47]: They are still relevant since we're not yet testing logs

File: test.js
Comment (juliangruber) [reply test.js:49]: See comment above

**Pull request merged**
